src/GitHub.App/Services/PullRequestService.cs

```diff
@@ -74,6 +84,156 @@ public IObservable<string> GetPullRequestTemplate(ILocalRepositoryModel reposito
+            {
+                var repo = gitService.GetRepository(repository.LocalPath);
+                var refspec = string.Format(CultureInfo.InvariantCulture, "{0}:{0}", repo.Head.FriendlyName);
+                return gitClient.Fetch(repo, "origin", new[] { refspec }).ToObservable();
+            });
+        }
+
+        public IObservable<Unit> FetchAndCheckout(ILocalRepositoryModel repository, int pullRequestNumber, string localBranchName)
+        {
+            return DoFetchAndCheckout(repository, pullRequestNumber, localBranchName).ToObservable();
+        }
+
+        public IObservable<string> GetDefaultLocalBranchName(ILocalRepositoryModel repository, int pullRequestNumber, string pullRequestTitle)
+        {
+            return Observable.Defer(() =>
+            {
+                var initial = "pr/" + pullRequestNumber + "-" + GetSafeBranchName(pullRequestTitle);
+                var current = initial;
+                var repo = gitService.GetRepository(repository.LocalPath);
+                var index = 2;
+
+                while (repo.Branches[current] != null)
+                {
+                    current = initial + '-' + index++;
+                }
+
+                return Observable.Return(current);
+            });
+        }
+
+        public IObservable<HistoryDivergence> CalculateHistoryDivergence(ILocalRepositoryModel repository, int pullRequestNumber)
+        {
+            return Observable.Defer(async () =>
+            {
+                var repo = gitService.GetRepository(repository.LocalPath);
+
+                EnsurePullRefSpecExists(repo);
+                await gitClient.Fetch(repo, "origin");
+
+                var pullRef = repo.Refs[$"refs/remotes/origin/pr/{pullRequestNumber}"] as DirectReference;
+                var commit = pullRef?.Target as Commit;
+
+                if (commit != null)
+                {
+                    var result = repo.ObjectDatabase.CalculateHistoryDivergence(repo.Head.Tip, commit);
+                    return Observable.Return(result);
+                }
+                else
+                {
+                    throw new InvalidOperationException("Could not find pull request ref.");
+                }
+            });
+        }
+
+        public IObservable<IBranch> GetLocalBranches(ILocalRepositoryModel repository, IPullRequestModel pullRequest)
+        {
+            return Observable.Defer(() =>
+            {
+                var repo = gitService.GetRepository(repository.LocalPath);
+                var result = GetLocalBranchesInternal(repository, repo, pullRequest).Select(x => new BranchModel(x, repository));
+                return result.ToObservable();
+            });
+        }
+
+        public bool IsPullRequestFromFork(ILocalRepositoryModel repository, IPullRequestModel pullRequest)
+        {
+            return pullRequest.Head.RepositoryCloneUrl.ToRepositoryUrl() != repository.CloneUrl.ToRepositoryUrl();
```

Comment:
> Ah, I didn't think of looking there. So if I check that do I still need to compare the clone URLs?

Yeah, I suppose you still do because it's possible that the PR target is the fork and not upstream.